[`base-controller`] Fix all `any` usage, apply universal supertype for functions

MajorLift opened the pull request:
## Motivation

The `ControllerMessenger` pattern that is used in all of our controllers currently relies on unsafe typing, which includes `any` usage. A more rigorous overhaul of these definitions will help us implement features with more type-safe and robust code. 

## Explanation

### Main

- Removes all `any` usage from `Action`, `Event` typing in `ControllerMessenger`. 
- Types `Action['handler']` with correct universal supertype for functions.
- Fixes `SelectorFunction`, `EventSubscriptionMap` typing that was causing errors. 
- Improves `BaseController` typing by defining `NamespacedName` type: the narrowest supertype of all names defined within a given namespace.
- `BaseController` now expects a "getState" action type and "stateChange" event type in all controllers.

### Auxiliary

- Improved typing for `deriveStateFromMetadata`
- ~Rename all single-letter type param names to be more descriptive.~ (moved here: https://github.com/MetaMask/core/pull/2030/)

### Notes on the "universal function supertype"

#### A. `(...args: never[]) => unknown`

`Action['handler'] extends ActionConstraint['handler']`, and in general `Subtype extends Supertype`. Therefore, in order to assign any function to `Action['handler']` without restrictions, we need `ActionConstraint['handler']` to be the widest possible function i.e. the supertype of all functions.

To achieve this, the return type of this supertype function needs to be `unknown` (universal supertype), and the parameters need to be `never` (universal subtype). This is because in general `(x: SuperType) => T` is a *subtype* of `(x: SubType) => T`. So the params need to be the narrowest type (i.e. `never`) for the function type to be widened into the universal function supertype. 

- There's an example of this in the [TypeScript docs](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types).
- And here's a more detailed [reddit comment](https://www.reddit.com/r/typescript/comments/muyl55/comment/gv9ndij/?utm_source=share&utm_medium=web2x&context=3) on this topic.

#### B. `(...args: never) => unknown`

In general, array types of arbitrary length are not assignable to tuple types that have fixed types designated to some or all of its index positions (the array is wider than the tuple since it has less constraints).

This means that there's a whole subcategory of functions that `(...args: never[]) => unknown` doesn't cover: functions that have a finite number of parameters that are strictly ordered and typed, making their spread argument list evaluate as a fixed-length or variadic **tuple** type, instead of a freely expandable (or contractable) **array** type.

That is, `(...args: never[]) => unknown` covers the cases where all of the params can be typed together as a group e.g. `(...args: (string | number)[]) => T`. But to cover cases like `(a: string, b: number, ...rest: unknown[]) => T`, we need another top type with an even narrower type for the params: `(...args: never) => unknown`. 

#### C. `((...args: never) => unknown) | ((...args: never[]) => unknown)`

In short, the above represents the top type that includes both of these function categories.

- `| never` in general is analogous to `+ 0`, so `(...args: (never[] | never)) => unknown` doesn't work.
- `(...args: never) => unknown` doesn't work with the error message `Type 'T[]' is not assignable to type 'never'`.

## References

- Closes https://github.com/MetaMask/core/issues/2026

## Changelog

### `@metamask/base-controller`
- **CHANGED**: `ActionConstraint['handler']` has been fixed to remove `any` usage, and is now defined as the universal supertype of all functions, meaning any function can be safely assigned to it, regardless of argument types, number of arguments, or return value type.
- **CHANGED** : `ExtractActionParameters` and `ExtractActionResponse` have also been fixed to remove `any` usage, and to use the new typing of `ActionConstraint['handler']`. 
- **CHANGED**: **BREAKING** Alters `SelectorFunction` type, replacing its `Args` generic parameter with an event payload type that is defined with a new required generic parameter `Event`.
- **ADDED**: Exports `NamespacedName` type, which is the narrowest supertype of all names defined within a given namespace.

### `@metamask/rate-limit-controller`
- **CHANGED**: **BREAKING** `RateLimitedApi['method']` is now constrained by the `ActionConstraint['handler']` type from the `@metamask/base-controller` module.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

BelfordZ commented on packages/queued-request-controller/src/QueuedRequestController.ts:
```diff
@@ -29,9 +41,23 @@ export type QueuedRequestControllerEvents =
+type AllowedActions =
+  | NetworkControllerGetStateAction
+  | NetworkControllerSetActiveNetworkAction
+  | NetworkControllerSetProviderTypeAction
```
this controller doesn't use any of these actions though.

MajorLift replied:
Is there a different name that you could suggest for this type union? I think we should avoid relying on copy-pastes to express this abstraction.

MajorLift replied:
@BelfordZ Would you mind taking a second look at this when you're free? 

Based on [`createQueuedRequestMiddleware`](https://github.com/MetaMask/core/pull/1890/files#diff-02c32f437b122b5c71f0ac6bdf304cd22f6d4569c9c9b8ecbb711c11c98cd0d3L51-L67) and [`buildMessenger`](https://github.com/MetaMask/core/pull/1890/files#diff-d85004a6e4bc8c032dffe450176f16ba4dad4806dce07bd2b2bb9838d73ca04aL23-L34), it looks like the messengers for `QueuedRequestController` need access to these `AllowedActions`, but I might be misunderstanding the context here.

MajorLift commented:
@mcmire Feel free to add suggestions or commits if any of the changes you mentioned might fit into this PR.

MajorLift commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -533,7 +565,7 @@ export class ControllerMessenger<
     }
 
-    subscribers.set(handler, selector);
+    subscribers.set(handler, selector as ExtractEventHandler<Event, E>);
```
Highlighting for closer review.

mcmire replied:
Hmm, will have to think about this some more.

MajorLift replied:
Updated this type here: https://github.com/MetaMask/core/pull/1890/commits/b1849805ac89323bc3f3460706281bb75f3ead46

MajorLift replied:
Continuing thread here: https://github.com/MetaMask/core/pull/1890#discussion_r1388479681

MajorLift commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -1,30 +1,46 @@
+> = Action extends {
   type: T;
-  handler: (...args: any) => infer H;
+  handler: (...args: never[]) => infer H extends unknown;
```
Highlighting to request closer review.

mcmire replied:
Hmm. Using `never[]` doesn't seem right here; we want to allow arguments, so I wonder if we ought to use `unknown[]` instead. Also, can you explain why we would need `extends unknown` here? I feel like this is redundant.

MajorLift replied:
Removed the `unknown` thanks for pointing that out!

MajorLift replied:
Regarding `never[]`, see: https://github.com/MetaMask/core/pull/1890#issuecomment-1783050539

MajorLift commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -41,13 +57,24 @@ export type SelectorEventHandler<SelectorReturnValue> = (
 export type ActionConstraint = {
   type: string;
-  handler: (...args: any) => unknown;
+  handler: (...args: never[]) => unknown;
```
Highlighting to request closer review.

mcmire replied:
Similar; this doesn't seem right. Does `unknown[]` work instead of `never[]`?

MajorLift replied:
This is a really fun one! So just in empirical terms, it's already clear that `never[]` is correct, both here and in `ExtractActionResponse`. Changing either to `unknown[]` (or even `[never]`) triggers a whole lot of errors all over the codebase, basically everywhere the `Action` type is used. 

But this is also logically sound, and the reason for this is interesting. 

Here we have `Action['handler'] extends ActionConstraint['handler']`, and in general `Subtype extends Supertype`. So we need `ActionConstraint['handler']` to be a function that is a supertype of all functions. 

To achieve this, the return type needs to be `unknown` (universal supertype), and the parameters need to be `never` (universal subtype). 
This is because while `number | string` is a supertype of `number`, `(x: number | string) => void` is a subtype of `(x: number) => void`.

- There's an example of this in the [TypeScript docs](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types)
- And here's a more detailed [reddit comment](https://www.reddit.com/r/typescript/comments/muyl55/comment/gv9ndij/?utm_source=share&utm_medium=web2x&context=3) on this topic.

MajorLift replied:
Regarding `never[]`, see: https://github.com/MetaMask/core/pull/1890#issuecomment-1783050539

MajorLift commented on packages/queued-request-controller/src/QueuedRequestMiddleware.test.ts:
```diff
@@ -1,44 +1,26 @@
   messenger: ReturnType<typeof buildMessenger>,
-  mocks: {
-    getNetworkClientById?: NetworkController['getNetworkClientById'];
+  mocks: ActionHandlerMap<
+    NetworkControllerGetNetworkClientByIdAction | AddApprovalRequest
+  > & {
```
Highlighting to request closer review.

MajorLift commented on packages/rate-limit-controller/src/RateLimitController.ts:
```diff
@@ -121,11 +129,15 @@ export class RateLimitController<
         ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]['method']>
-      ) => this.call(origin, type, ...args)) as any,
+      ) =>
+        this.call(origin, type, ...args) as ExtractActionResponse<
+          CallApi<RateLimitedApis>,
+          'RateLimitController:call'
+        >,
```
Highlighting to request closer review.

MajorLift commented on packages/rate-limit-controller/src/RateLimitController.ts:
```diff
@@ -155,7 +167,9 @@ export class RateLimitController<
     }
 
-    return implementation(...args);
+    return implementation(...args) as ReturnType<
+      RateLimitedApis[ApiType]['method']
+    >;
```
Highlighting to request closer review.

mcmire commented on packages/base-controller/src/BaseControllerV2.ts:
```diff
@@ -65,7 +67,13 @@ export interface StatePropertyMetadata<T extends Json> {
   N extends string,
   S extends Record<string, Json>,
-  messenger extends RestrictedControllerMessenger<N, any, any, string, string>,
+  messenger extends RestrictedControllerMessenger<
```
This is interesting. This would be a breaking change, and so I am wary of changing this. That said, I don't understand why we chose to use `any` here. I would be curious to hear if Mark can remember any context around this.

MajorLift replied:
`ActionConstraint` and `EventConstraint` in this PR are written to be the universal supertypes of their respective categories, so I don't think we'd be adding a meaningful constraint by moving away from `any`. Also no tests are being broken by this change, although I guess we'll have to check whether any of the other repos are using incompatible types.

Gudahtt replied:
In general, using `any` in this context is not harmful in the same way that it is in other contexts. Types used in a generic constraints don't get applied to any specific variables, they're just constraints (i.e. by using `any` here, absolutely nothing is actually given the type `any`, even indirectly). For this reason I've sometimes put less effort into avoiding the use of `any` for generic constraints.

That said, more detailed constraints don't hurt either. And in this case, these constraints are already present in the `RestrictedControllerMessenger` type so this change seems like it would not affect anything functionally.

MajorLift replied:
Generic constraints are also helpful for intellisense hints and self-documenting code. If finding the right constraints isn't a significant blocker for actual feature development, I think it's worth the extra effort, even if it's not strictly necessary in terms of functionality.

mcmire commented on packages/base-controller/src/BaseControllerV2.ts:
```diff
@@ -247,29 +255,30 @@ function deriveStateFromMetadata<S extends Record<string, Json>>(
-      const stateMetadata = metadata[key as keyof S];
-      if (!stateMetadata) {
-        throw new Error(`No metadata found for '${key}'`);
+  return Object.keys(state).reduce(
```
Nice 👍🏻

mcmire commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -1,30 +1,46 @@
-export type ExtractActionParameters<Action, T> = Action extends {
+
+export type ExtractActionParameters<
+  Action extends ActionConstraint,
+  T = Action['type'],
+> = Action extends {
   type: T;
-  handler: (...args: infer H) => any;
+  handler: (...args: infer H) => unknown;
 }
   ? H
   : never;
```
Hmm. We already seem to be checking that `Action` extends `ActionConstraint` by virtue of the conditional type. Does it makes sense to restrict what it needs to be in the type parameters too?

Minimally it seems like we ought to drop `T`, since that's all we really need `Action` to have. We also ought to use a more than one letter type parameter:

``` suggestion
export type ExtractActionParameters<Action> = Action extends {
  handler: (...args: infer Handler) => unknown;
}
  ? Handler
  : never;
```

The same argument applies to all of the types in this section (I won't add comments for them, but I may come back to this).

MajorLift replied:
I agree that dropping `T` is probably the way to go with this. I wanted to minimize disruption in case the second parameter was being used somewhere. I'll check whether we can remove the `T` params from all of the types here.

Gudahtt replied:
Agreed except that `unknown` would be the wrong return type here; we'd still want `any`. We want this to be compatibile with handlers that return any type, not just `unknown`.

Gudahtt replied:
Ah right, makes sense, thanks

MajorLift replied:
The `{Action,Event}Type` params seem to be necessary for type inference to work around action/event types and the methods that accept or return them. Removing just the event types still triggers "implicit any" errors across multiple controllers.

mcmire replied:
Okay, no problem.

mcmire reviewed (commented):
I left some comments on the `base-controller` changes that I was curious about. What do you think about splitting those off into a separate PR? I feel like they could be high impact and so it'd be better to have them in their own PR so they're easier to review now and later.

MajorLift commented:
@mcmire Thanks for the detailed comments! I'll split off the `BaseController` changes into a separate PR as soon as the related comments here are resolved.

MajorLift commented:
@mcmire On the subject of `(...args: never[]) => unknown`: 
- https://github.com/MetaMask/core/pull/1890#discussion_r1373781696
- https://github.com/MetaMask/core/pull/1890#discussion_r1373782754

This is a really fun one! `never[]` is actually the logically correct type here, as unintuitive as that seems.

So just in empirical terms, changing the `never[]` to `unknown[]` (or even `[never]`) triggers a whole lot of errors all over the codebase, basically everywhere the `Action` type is used. 

Here's why: `Action['handler'] extends ActionConstraint['handler']`, and in general `Subtype extends Supertype`. Therefore, we need `ActionConstraint['handler']` to be a function that is a supertype of all functions, so we can freely assign any function to `Action['handler']`.

To achieve this, the return type of this supertype function needs to be `unknown` (universal supertype), and the parameters need to be `never` (universal subtype). This is because while `number | string` is a supertype of `number`, `(x: number | string) => void` is a *subtype* of `(x: number) => void`.

- There's an example of this in the [TypeScript docs](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#inferring-within-conditional-types).
- And here's a more detailed [reddit comment](https://www.reddit.com/r/typescript/comments/muyl55/comment/gv9ndij/?utm_source=share&utm_medium=web2x&context=3) on this topic.

FrederikBolding commented:
Should this not be separate PRs to improve reviewability?

MajorLift commented:
@FrederikBolding Agreed! I'll try and split this up into smaller PRs.

MajorLift commented:
Pivoting to focus this PR on the `BaseController`, `RateLimitController` type fixes, since the discussion here is concentrated on those changes.

Will open separate PR for `QueuedRequestController`, `GasController` type fixes.

MajorLift commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -1,32 +1,48 @@
-  handler: (...args: infer H) => any;
+
+export type ExtractActionParameters<
+  Action extends ActionConstraint,
+  ActionType = Action['type'],
+> = Action extends {
+  type: ActionType;
+  handler: (...args: infer HandlerArgs) => unknown;
 }
-  ? H
+  ? HandlerArgs
```
Continuing this thread: https://github.com/MetaMask/core/pull/1890#discussion_r1373773411

Mrtenz commented on packages/base-controller/src/BaseControllerV2.ts:
```diff
@@ -68,7 +70,13 @@ export interface StatePropertyMetadata<T extends Json> {
   S extends Record<string, Json>,
-  messenger extends RestrictedControllerMessenger<N, any, any, string, string>,
+  messenger extends RestrictedControllerMessenger<
+    N,
```
If we're doing a refactor, I would prefer moving away from non-descriptive single-letter generics. I have no idea what `N` and `S` are without looking at more of the source code currently.

MajorLift replied:
Good call! I got rid of all the remaining single-letter generics here: https://github.com/MetaMask/core/pull/1890/commits/08413a9ed9b16d92e4c782281f99b45aa4ef71c9

MajorLift replied:
Extracted to new PR: https://github.com/MetaMask/core/pull/2030/

Mrtenz commented on packages/base-controller/src/BaseControllerV2.ts:
```diff
@@ -117,7 +125,7 @@ export class BaseController<
 
     this.messagingSystem.registerActionHandler(
-      `${name}:getState`,
+      String(`${name}:getState`),
```
What's the reasoning behind this?

MajorLift replied:
Replacing with a pattern used elsewhere in the file for now. I'll revisit the `any` usage here. 

```suggestion
      `${name}:getState` as Namespaced<ControllerName, any>,
```

MajorLift replied:
All of the `Namespaced` with `any` usage here has to do with the awkward way the `Namespaced` type is defined. I'm looking into ways we could do this differently.

MajorLift replied:
Addressed: ce08095bd1de347aa43046bfb6024348356ac8e8 and ae291da004b86a2984448ca0d71def73ba0ac97f

Mrtenz commented on packages/base-controller/src/BaseControllerV2.ts:
```diff
@@ -250,29 +258,30 @@ function deriveStateFromMetadata<S extends Record<string, Json>>(
+        }
+        return persistedState;
+      } catch (error) {
+        // Throw error after timeout so that it is captured as a console error
+        // (and by Sentry) without interrupting state-related operations
+        setTimeout(() => {
+          throw error;
+        });
```
I know this was originally here, but this seems very hacky 😅 Is there no other way to capture an error in Sentry, without throwing?

Gudahtt replied:
We could accept a `captureError` method in the constructor, but IMO this solution is a lot nicer. No need for extra required constructor params for every single controller. Relying on global error handlers is framework-agnostic too.

We could hide this behind a utility function. Not sure we'll need it in very many places though.

Mrtenz commented on packages/base-controller/src/BaseControllerV2.ts:
```diff
@@ -250,29 +258,30 @@ function deriveStateFromMetadata<S extends Record<string, Json>>(
+      try {
+        const stateMetadata = metadata[key];
+        if (!stateMetadata) {
+          throw new Error(`No metadata found for '${String(key)}'`);
+        }
+        const propertyMetadata = stateMetadata[metadataProperty];
+        const stateProperty = state[key];
+        if (typeof propertyMetadata === 'function') {
+          persistedState[String(key)] = propertyMetadata(stateProperty);
+        } else if (propertyMetadata) {
+          persistedState[String(key)] = stateProperty;
```
Why all the `String`s?

MajorLift replied:
Figured out a better way to handle this: b32a7ed537591e8fa0a63786d09f327e0a632a7f

MajorLift replied:
This is fundamentally because `keyof Record<string, ...>` is `string | number | symbol`.

MajorLift replied:
Further updated here: ee2467dafecd8ca83c736589ff9462e74421ee38
See: https://github.com/MetaMask/core/pull/1890#discussion_r1390574901

Mrtenz commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -533,7 +552,7 @@ export class ControllerMessenger<
     }
 
-    subscribers.set(handler, selector);
+    subscribers.set(handler, selector as SelectorFunction<unknown[], V>);
```
Why does this need to be cast?

MajorLift replied:
We need better typing for `EventSubscriptionMap`. Working on this.

MajorLift replied:
Type casting removed and `EventSubscriptionMap` typing improved here: 7baebd7814ea5f65962cb4316825c50d2488bfd6

Mrtenz commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -154,7 +173,7 @@ export class RestrictedControllerMessenger<
    * @template T - A type union of Action type strings that are namespaced by N.
    */
-  registerActionHandler<T extends Namespaced<N, Action['type']>>(
+  registerActionHandler<T extends Action['type']>(
```
Single-letter generic.

MajorLift replied:
Resolved here: https://github.com/MetaMask/core/pull/1890/commits/08413a9ed9b16d92e4c782281f99b45aa4ef71c9

Mrtenz commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -154,7 +173,7 @@ export class RestrictedControllerMessenger<
    * @throws Will throw when a handler has been registered for this action type already.
    * @template T - A type union of Action type strings that are namespaced by N.
```
This comment seems out of date now.

MajorLift replied:
Fixed here f1cc1236578ebc263cb14b44a04c4f6027938801

MajorLift replied:
Good catch! Fixed here: https://github.com/MetaMask/core/pull/1890/commits/f1cc1236578ebc263cb14b44a04c4f6027938801

Mrtenz commented on packages/rate-limit-controller/src/RateLimitController.ts:
```diff
@@ -121,11 +129,14 @@ export class RateLimitController<
         ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]['method']>
-      ) => this.call(origin, type, ...args)) as any,
+      ) =>
+        this.call(origin, type, ...args) as ExtractActionResponse<
+          CallApi<RateLimitedApis>
+        >,
```
Can this cast maybe be replaced with a generic type on the `call` method?

MajorLift replied:
See https://github.com/MetaMask/core/pull/1890/files#r1388483659

MajorLift replied:
Narrowing the `call` method generic param `RateLimitedApis` doesn't help here, since the `CallApi` action handler being passed into `.registerActionHandler()` needs to work for all `RateLimitedApis` defined in the generic argument that instantiates the `RateLimitController` class.

I don't think a type cast can be avoided here, since TypeScript is not inferring that `ReturnType<this.call>` and `ExtractActionResponse<CallApi<RateLimitedApis>>` are equivalent by definition. The type cast itself is essentially a tautology, and therefore shouldn't introduce typing issues.

MajorLift replied:
A breakdown of the confusing type relationships here:
- `CallApi<RateLimitedApis>['handler'] extends RateLimitedApis[ApiType]['method']`
- `RateLimitController<RateLimitedApis>['call'] extends RateLimitedApis[ApiType]['method']`
- `CallApi<RateLimitedApis>['handler'] := RateLimitController<RateLimitedApis>['call']`

MajorLift replied:
Type cast removed here: 01eb2a002b1cdd32714554f43afde9534cfe286e

Mrtenz commented on packages/rate-limit-controller/src/RateLimitController.ts:
```diff
@@ -149,13 +160,14 @@ export class RateLimitController<
     }
 
-    return implementation(...args);
+    return implementation(...args) as ReturnType<typeof implementation>;
```
Why are we casting to the return type of the same function? Seems odd 😅

MajorLift replied:
This and line 137 are both caused by the contradictory typing of `RateLimitedApis[ApiType]['method']`, specifically how it's being used simultaneously as the widest type constraint and as the narrow method subtypes for individual `ApiType`s. 

I'll see if there's a way to reconcile this. Hopefully both type casts can be removed if the underlying issue is resolved.

MajorLift replied:
Tentatively addressed here: https://github.com/MetaMask/core/pull/1890#discussion_r1391340917

MajorLift replied:
Improved here: e8826efdf87ec1d65122f8809eff07f191d0967b

MajorLift replied:
See https://github.com/MetaMask/core/pull/1890#discussion_r1392026749

MajorLift replied:
Fixed: e8826efdf87ec1d65122f8809eff07f191d0967b

Gudahtt commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -513,27 +550,36 @@ export class ControllerMessenger<
-    subscribers.set(handler, selector);
+    subscribers.set(
+      handler,
+      selector as SelectorFunction<unknown[], SelectorReturnValue>,
```
I see that this cast was needed because of the change to `EventSubscriptionMap`. I am having trouble understanding why this is an error though.

MajorLift replied:
Type casting removed and `EventSubscriptionMap` typing improved here: 7baebd7814ea5f65962cb4316825c50d2488bfd6.

`EventSubscriptionMap` (`this.events` subscription map) was typed as its category supertype, which made it impossible to assign selector function instances with narrower payloads than `unknown[]` to `this.events`.

Gudahtt commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -41,37 +59,41 @@ export type SelectorEventHandler<SelectorReturnValue> = (
 export type ActionConstraint = {
   type: string;
-  handler: (...args: any) => unknown;
+  handler: (...args: never[]) => unknown;
```
This change seems to be responsible for breaking the types in the rate limit controller. If we remove this change, the rate limit controller changes can be removed as well, and this becomes a non-breaking PR.

Gudahtt replied:
Hmm.... it seems to be non-breaking if we make it `never` though, rather than `never[]`. At the moment it's typed as `any`; even changing it to `any[]` introduces type errors.

MajorLift replied:
Addressed here: d474b253032f098605f32321246a5e736605dd9c. I think this makes `ActionConstraint` non-breaking?

I tentatively typed `ReturnType<RateLimitedApi['method']>` as `boolean` based on the specified [jsdoc return type](https://github.com/MetaMask/core/pull/1890/commits/d474b253032f098605f32321246a5e736605dd9c#diff-1eb134c9c5a9dd0a4e4838f719ba67723f61fdacd8bfc1a9acef96366b7578fbR149), but replacing the `boolean` with `unknown` also works.

Gudahtt replied:
Hmm, it still seems like there are breaking changes here to the base controller package. I tried undoing the `rate-limit-controller` changes and there are two type errors:

```
packages/rate-limit-controller/src/RateLimitController.ts:57:3 - error TS2344: Type 'RateLimitControllerActions<RateLimitedApis>' does not satisfy the constraint 'ActionConstraint'.
  Type 'CallApi<RateLimitedApis>' is not assignable to type 'ActionConstraint'.
    Types of property 'handler' are incompatible.                                                                                                                                          
      Type '<ApiType extends keyof RateLimitedApis>(origin: string, type: ApiType, ...args: Parameters<RateLimitedApis[ApiType]["method"]>) => Promise<...>' is not assignable to type '(..
.args: never[]) => unknown'.                                                                                                                                                                       Types of parameters 'origin' and 'args' are incompatible.
          Type 'never[]' is not assignable to type '[origin: string, type: never, ...args: Parameters<RateLimitedApis[never]["method"]>]'.
            Target requires 3 element(s) but source may have fewer.                                                                                                                        

57   RateLimitControllerActions<RateLimitedApis>,                                                                                                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~   

packages/rate-limit-controller/src/RateLimitController.ts:75:3 - error TS2344: Type 'RateLimitMessenger<RateLimitedApis>' does not satisfy the constraint 'RestrictedControllerMessenger<"R
ateLimitController", ActionConstraint | ControllerActions<"RateLimitController", RateLimitState<RateLimitedApis>>, EventConstraint | ControllerEvents<...>, string, string>'.
  Types of property 'registerActionHandler' are incompatible.
    Type '<ActionType extends "RateLimitController:getState" | "RateLimitController:call">(action: ActionType, handler: ActionHandler<RateLimitControllerActions<RateLimitedApis>, ActionTy
pe>) => void' is not assignable to type '<ActionType extends string>(action: ActionType, handler: ActionHandler<ActionConstraint | ControllerActions<"RateLimitController", RateLimitState<
RateLimitedApis>>, ActionType>) => void'.
      Types of parameters 'handler' and 'handler' are incompatible.      
        Types of parameters 'args' and 'args' are incompatible.
          Type '[] | [origin: string, type: keyof RateLimitedApis, ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]["method"]>]' is not assignable to type 'never[] | []'.
            Type '[origin: string, type: keyof RateLimitedApis, ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]["method"]>]' is not assignable to type 'never[] | []'.
              Type '[origin: string, type: keyof RateLimitedApis, ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]["method"]>]' is not assignable to type 'never[]'.
                Type 'string | keyof RateLimitedApis | Parameters<RateLimitedApis[keyof RateLimitedApis]["method"]>[number]' is not assignable to type 'never'.
                  Type 'string' is not assignable to type 'never'.

75   RateLimitMessenger<RateLimitedApis>

```

MajorLift replied:
This is caused by the removal of `any` in `ActionContraint['handler']` itself and in the generic params for ControllerMessenger. 

The change that resolves this issue is at [lines 52-54 of rate=limit-controller](https://github.com/MetaMask/core/pull/1890/files/1fa24a826e4bec9c1e9e01aea8dbb3105438528a#diff-1eb134c9c5a9dd0a4e4838f719ba67723f61fdacd8bfc1a9acef96366b7578fbR52-R54).

In all of these cases, no actual constraints are being added, since `ActionConstraint['handler']` is the supertype of all functions, but since previously suppressed errors are popping up, I added a **BREAKING** notice in the changelog for `ActionConstraint` and `EventConstraint`.

mcmire replied:
Ah, so it sounds like the reason why this breaks `rate-limit-controller` is that the types there were incorrect. Is this true?

MajorLift replied:
Yes the constraint in `RestrictedControllerMessenger` enforcing that `CallApi extends ActionConstraint` was already in place -- it was just being suppressed by the `any` usage (the definition of `this.call()` and `CallApi` both derive from `RateLimitedApi['method']`, which was previously typed with `any`).

MajorLift commented on packages/base-controller/src/BaseControllerV2.ts:
```diff
@@ -245,34 +259,38 @@ export function getPersistentState<S extends Record<string, Json>>(
-        throw new Error(`No metadata found for '${key}'`);
+  return Object.keys(state).reduce(
+    (
+      persistedState: Record<keyof ControllerState, Json>,
```
Applied here: ee2467dafecd8ca83c736589ff9462e74421ee38.

This seems like a fundamental limitation of `reduce`, as it doesn't distinguish between the initial value type and the eventual return type. It might be worth writing a utility type for typing reduce that lets us specify both, with the initial value type defaulting to `Partial<ReturnValueType>`.

`key` resolving to `string` is probably because the return type of `Object.keys()` defaults to `string[]` without further inference-based narrowing. This casting gives us good intellisense/autocomplete: `Object.keys(state) as (keyof ControllerState)[]`.

MajorLift commented on packages/base-controller/src/BaseControllerV2.ts:
```diff
@@ -107,17 +115,17 @@ export class BaseController<
 
     this.messagingSystem.registerActionHandler(
-      `${name}:getState`,
+      `${name}:getState` as Namespaced<ControllerName, any>,
```
Addressed: ce08095bd1de347aa43046bfb6024348356ac8e8 and ae291da004b86a2984448ca0d71def73ba0ac97f

MajorLift replied:
It looks like we could replace all of the "getState" and "stateChange" types across the codebase with the polymorphic `ControllerGetStateAction` and `ControllerStateChangeEvent` now defined in base-controller e.g. e21051809bdba24ccda8da80cb0146fa6bde1f4b.

MajorLift replied:
Extracted here: https://github.com/MetaMask/core/pull/2029

mcmire reviewed (commented):
I've done a final pass on this PR. I think this reveals that we need to take another look at the types in ControllerMessenger at some point to see if we can simplify them, because they are extremely complicated now and I feel like they shouldn't be. I also keep getting confused on what the difference between Action and AllowedAction is (or Event vs. AllowedEvent).

In any case, I just had two more questions, and then I'm good on this.

MajorLift commented on packages/assets-controllers/src/CurrencyRateController.ts:
```diff
@@ -35,24 +38,26 @@ export type CurrencyRateState = {
-  CurrencyRateStateChange,
+  CurrencyRateControllerActions | AllowedActions,
+  CurrencyRateControllerEvents,
   AllowedActions['type'],
```
I reverted the codebase-wide controller refactors to extract them to a separate PR, but the `CurrencyRateControllerActions` union here contains `GetCurrencyRateState` so it's still being passed in as a generic argument. 

`AllowedAction` defines usable actions that are imported from other controllers, and `${typeof name}Actions` enumerates available actions that are internal to the controller. So in general, the generic arguments used here are how a `ControllerMessenger` should be initiated.

MajorLift replied:
PR extracted here: https://github.com/MetaMask/core/pull/2029

MajorLift commented on packages/gas-fee-controller/src/GasFeeController.ts:
```diff
@@ -216,26 +219,27 @@ export type GasFeeState = GasFeeEstimatesByChainId & SingleChainGasFeeState;
+  typeof name,
+  GasFeeControllerActions | AllowedActions,
+  GasFeeControllerEvents | NetworkControllerStateChangeEvent,
+  AllowedActions['type'],
```
Same as above https://github.com/MetaMask/core/pull/1890#discussion_r1391491458

MajorLift commented on packages/rate-limit-controller/src/RateLimitController.ts:
```diff
@@ -141,7 +152,7 @@ export class RateLimitController<
     type: ApiType,
     ...args: Parameters<RateLimitedApis[ApiType]['method']>
-  ): Promise<ReturnType<RateLimitedApis[ApiType]['method']>> {
+  ): Promise<boolean> {
```
Highlighting for closer review: should the `call` method here return `boolean`?

MajorLift replied:
Resolved here: https://github.com/MetaMask/core/pull/1890#discussion_r1391739206

MajorLift replied:
See https://github.com/MetaMask/core/pull/1890#discussion_r1392000743

MajorLift replied:
Fixed here: 33d23286a871b80cc9bfe1022744312ad9889744

- Removes `@returns` jsdoc tag for `call` method that specified its return type as `boolean`.
- Fixes implicit `any` typing for `implementation`.
- Removes now unnecessary type casting of `this.call()` in `registerActionHandler` message for `call` method.

MajorLift commented on packages/rate-limit-controller/src/RateLimitController.ts:
```diff
@@ -10,7 +14,7 @@ import type { Patch } from 'immer';
  */
 export type RateLimitedApi = {
-  method: (...args: any[]) => any;
+  method: (...args: never[]) => boolean;
```
Highlighting for closer review: What is the expected shape of a rate-limited method? Should the return type here be `unknown`?

mcmire replied:
Yes, probably. I don't think `boolean` makes sense here. These methods could in theory take anything and return anything. Because of this, the arguments and return value should probably be generics so that what `call` returns matches the return type of the method that it's calling. That feels like a larger change to me, though...

MajorLift replied:
Reverted here: e8826efdf87ec1d65122f8809eff07f191d0967b
This commit also casts the return type of `call` to the return type of the method being called.

MajorLift replied:
Fixed here: 33d23286a871b80cc9bfe1022744312ad9889744

- Removes erroneous `@returns` jsdoc tag for `call` method that specifies its return type as `boolean`.
- Fixes implicit `any` typing for `implementation`.
- Removes unnecessary type casting of `this.call()` in `registerActionHandler` message for `call` method.

MajorLift replied:
See https://github.com/MetaMask/core/pull/1890#discussion_r1392026749

Gudahtt commented:
It might be better in the future to separate noisy changes like variable renames from other refactors. Most of the diff for this PR is variable renames. The few actual type changes are easy to miss, and would be much easier to review if they were separate

MajorLift commented:
> It might be better in the future to separate noisy changes like variable renames from other refactors. Most of the diff for this PR is variable renames. The few actual type changes are easy to miss, and would be much easier to review if they were separate

Renaming single-letter type params was requested by the other reviewers, but I agree that any readability benefits are outweighed by the distraction caused by the noisy diffs. 

I extracted the renames to a separate PR: https://github.com/MetaMask/core/pull/2030/ and reverted those changes here.

MajorLift commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -1,38 +1,58 @@
 
-export type SelectorFunction<Args extends unknown[], ReturnValue> = (
-  ...args: Args
+export type SelectorFunction<Event extends EventConstraint, ReturnValue> = (
+  ...args: ExtractEventPayload<Event>
```
Highlighting for closer review

Gudahtt replied:
Good catch, that does seem like a much more appropriate type

mcmire reviewed (dismissed):
There are clearly opportunities for us to keep improving types in RateLimitController — I suspect we may not be doing things the best way there. But overall I'm good with these changes.

Gudahtt commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -151,10 +180,10 @@ export class RestrictedControllerMessenger<
    * @template T - A type union of Action type strings that are namespaced by N.
    */
-  registerActionHandler<T extends Namespaced<N, Action['type']>>(
+  registerActionHandler<T extends Action['type']>(
```
It looks like this type no longer requires the action to be in the current namespace. Was that intentional?

Gudahtt replied:
e.g. maybe it should be

```suggestion
  registerActionHandler<T extends NamespacedName<N>>(
```

MajorLift replied:
Oh great catch! I forgot to restore this constraint. Fixed: d84ade9ffa9ca59311155b1876fdf921097815cc

Gudahtt reviewed (dismissed):
LGTM!

mcmire reviewed (approved):
Looks good!

mcmire commented:
@Mrtenz Do you have any more concerns about this PR?

MajorLift commented on packages/base-controller/src/ControllerMessenger.ts:
```diff
@@ -41,13 +61,19 @@ export type SelectorEventHandler<SelectorReturnValue> = (
 
 export type ActionConstraint = {
   type: string;
-  handler: (...args: any) => unknown;
+  handler: ((...args: never) => unknown) | ((...args: never[]) => unknown);
+};
```
So it turns out that `(...args: never[]) => unknown` probably isn't the universal supertype for functions.

The error messages @Gudahtt linked here https://github.com/MetaMask/core/pull/1890#discussion_r1391617360 reflect the fact that in general, array types of arbitrary length are not assignable to tuple types that have fixed types designated to each index position (since the array is wider than the tuple).

This means that there's a whole subcategory of functions that `(...args: never[]) => unknown` doesn't cover: functions that have a finite number of parameters that are strictly ordered and typed, making their argument list evaluate as a fixed-length or variadic **tuple** type, instead of an expandable **array** type.

That is, `(...args: never[]) => unknown` covers the cases where all of the params can be typed together as a group e.g. `(...args: (string | number)[]) => T`. But to cover cases like `(a: string, b: number, ...rest: unknown[]) => T`, we need another top type with an even narrower type for the params: `(...args: never) => unknown`. 

In short, I think the top type that includes both of these function categories is: `((...args: never) => unknown) | ((...args: never[]) => unknown)`.

(`| never` in general is analogous to `+ 0`, so `(...args: (never[] | never)) => unknown` doesn't work.)

mcmire replied:
Great sleuthing! 👏🏻

mcmire reviewed (approved):
Looks great!

Mrtenz reviewed (approved):
Looks great!

Gudahtt commented:
Could you update the PR description to clarify that the action constraint change is no longer breaking? I think with this most recent update that should be the case

MajorLift commented:
@Gudahtt Changelog updated!
